### ISSUE TITLE
Appium 1.3.4 fails to start when using robotium backend on Android Real Device

### ISSUE THREAD

**Issue by aluedeke** [appium/appium#4414]
Hey,

i am trying to use Appium with the Robotium backend but the session creation always fails with the following log. I am using a Lollipop real device for testing on an linux host. Any ideas how to solve this:

```
info: Client User-Agent string: Apache-HttpClient/4.3.4 (java 1.5)
info: [debug] Using local app from desired caps: /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk
info: [debug] Creating new appium session 4e9f54e9-9289-408f-9d1c-f7ab41d24a7c
info: [debug] Starting selendroid server
info: [debug] Getting Java version
info: Java version is: 1.8.0_25
info: [debug] Checking whether selendroid is built yet
info: [debug] Selendroid server exists!
info: [debug] Using fast reset? true
info: [debug] Preparing device for session
info: [debug] Checking whether app is actually present
info: [debug] Checking whether adb is present
info: [debug] Using adb from /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb
info: Retrieving device
info: [debug] Trying to find a connected android device
info: [debug] Getting connected devices...
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb devices
info: [debug] 1 device(s) connected
info: Found device 07217e2423b790db
info: [debug] Setting device id to 07217e2423b790db
info: [debug] Waiting for device to be ready and to respond to shell commands (timeout = 5)
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db wait-for-device
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "echo 'ready'"
info: [debug] Starting logcat capture
info: [debug] Checking whether aapt is present
info: [debug] Using aapt from /home/aluedeke/Development/tools/android-sdk-linux/build-tools/21.1.2/aapt
info: [debug] Checking if has internet permission from manifest.
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/build-tools/21.1.2/aapt dump badging /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk
info: [debug] Rebuilt selendroid server already exists, no need to rebuild it with a new manifest
info: [debug] Checking signed status of /tmp/selendroid.com.getyourguide.android.apk
info: [debug] Checking app cert for /tmp/selendroid.com.getyourguide.android.apk.
info: [debug] executing cmd: java -jar /home/aluedeke/Development/source/appium/gyg/node_modules/appium/node_modules/appium-adb/jars/verify.jar /tmp/selendroid.com.getyourguide.android.apk
info: [debug] Checking signed status of /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk
info: [debug] Checking app cert for /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk.
info: [debug] executing cmd: java -jar /home/aluedeke/Development/source/appium/gyg/node_modules/appium/node_modules/appium-adb/jars/verify.jar /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk
info: [debug] App already signed.
info: [debug] Zip-aligning /tmp/selendroid.com.getyourguide.android.apk
info: [debug] Checking whether zipalign is present
info: [debug] Using zipalign from /home/aluedeke/Development/tools/android-sdk-linux/build-tools/21.1.2/zipalign
info: [debug] Zip-aligning apk.
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/build-tools/21.1.2/zipalign -f 4 /tmp/selendroid.com.getyourguide.android.apk /tmp/appium115020-27744-wgklmw.tmp
info: [debug] App already signed.
info: [debug] Zip-aligning /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk
info: [debug] Checking whether zipalign is present
info: [debug] Using zipalign from /home/aluedeke/Development/tools/android-sdk-linux/build-tools/21.1.2/zipalign
info: [debug] Zip-aligning apk.
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/build-tools/21.1.2/zipalign -f 4 /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk /tmp/appium115020-27744-179n00i.tmp
info: [debug] Rebuilt selendroid apk does not exist, uninstalling any instances of it on device to make way for new one
info: [debug] Uninstalling com.getyourguide.android.selendroid
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "am force-stop com.getyourguide.android.selendroid"
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db uninstall com.getyourguide.android.selendroid
info: [debug] App was not uninstalled, maybe it wasn't on device?
info: [debug] Getting install status for com.getyourguide.android.selendroid
info: [debug] Getting device API level
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "getprop ro.build.version.sdk"
info: [debug] Device is at API Level 21
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "pm list packages -3 com.getyourguide.android.selendroid"
info: [debug] App is not installed
info: [debug] Rebuilt selendroid is not installed, installing it
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db install -r "/tmp/selendroid.com.getyourguide.android.apk"
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "getprop persist.sys.language"
info: [debug] Current device persist.sys.language: en
info: [debug] java -jar "/home/aluedeke/Development/source/appium/gyg/node_modules/appium/node_modules/appium-adb/jars/appium_apk_tools.jar" "stringsFromApk" "/home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk" "/tmp/com.getyourguide.android" en
info: [debug] No strings.xml for language 'en', getting default strings.xml
info: [debug] java -jar "/home/aluedeke/Development/source/appium/gyg/node_modules/appium/node_modules/appium-adb/jars/appium_apk_tools.jar" "stringsFromApk" "/home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk" "/tmp/com.getyourguide.android"
info: [debug] Not uninstalling app since server not started with --full-reset
info: [debug] Checking app cert for /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk.
info: [debug] executing cmd: java -jar /home/aluedeke/Development/source/appium/gyg/node_modules/appium/node_modules/appium-adb/jars/verify.jar /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk
info: [debug] App already signed.
info: [debug] Zip-aligning /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk
info: [debug] Checking whether zipalign is present
info: [debug] Using zipalign from /home/aluedeke/Development/tools/android-sdk-linux/build-tools/21.1.2/zipalign
info: [debug] Zip-aligning apk.
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/build-tools/21.1.2/zipalign -f 4 /home/aluedeke/Development/source/appium/gyg/src/test/resources/getyourguide-playstore.apk /tmp/appium115020-27744-9oaqya.tmp
info: [debug] MD5 for app is 4bd485b6463458102644ebe55ffb6466
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "ls /data/local/tmp/4bd485b6463458102644ebe55ffb6466.apk"
info: [debug] Getting install status for com.getyourguide.android
info: [debug] Getting device API level
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "getprop ro.build.version.sdk"
info: [debug] Device is at API Level 21
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "pm list packages -3 com.getyourguide.android"
info: [debug] App is installed
info: App is already installed, resetting app
info: [debug] Running fast reset (stop and clear)
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "am force-stop com.getyourguide.android"
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "pm clear com.getyourguide.android"
info: [debug] Forwarding system:8080 to device:8080
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db forward tcp:8080 tcp:8080
info: [debug] Pushing settings apk to device...
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db install "/home/aluedeke/Development/source/appium/gyg/node_modules/appium/build/settings_apk/settings_apk-debug.apk"
info: [debug] Pushing unlock helper app to device...
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db install "/home/aluedeke/Development/source/appium/gyg/node_modules/appium/build/unlock_apk/unlock_apk-debug.apk"
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "dumpsys window"
info: [debug] Screen already unlocked, continuing.
info: [debug] executing cmd: /home/aluedeke/Development/tools/android-sdk-linux/platform-tools/adb -s 07217e2423b790db shell "am instrument -e main_activity 'com.getyourguide.android/.activities.HomeActivity' com.getyourguide.android.selendroid/io.selendroid.ServerInstrumentation"
error: android.util.AndroidException: INSTRUMENTATION_FAILED: com.getyourguide.android.selendroid/io.selendroid.ServerInstrumentation
	at com.android.commands.am.Am.runInstrument(Am.java:951)
	at com.android.commands.am.Am.onRun(Am.java:316)
	at com.android.internal.os.BaseCommand.run(BaseCommand.java:47)
	at com.android.commands.am.Am.main(Am.java:99)
	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:249)
info: [debug] Cleaning up appium session
error: Failed to start an Appium session, err was: Error: android.util.AndroidException: INSTRUMENTATION_FAILED: com.getyourguide.android.selendroid/io.selendroid.ServerInstrumentation
info: [debug] Error: android.util.AndroidException: INSTRUMENTATION_FAILED: com.getyourguide.android.selendroid/io.selendroid.ServerInstrumentation
    at /home/aluedeke/Development/source/appium/gyg/node_modules/appium/node_modules/appium-adb/lib/adb.js:1378:17
    at null.<anonymous> (/home/aluedeke/Development/source/appium/gyg/node_modules/appium/node_modules/appium-adb/lib/adb.js:172:9)
    at ChildProcess.exithandler (child_process.js:646:7)
    at ChildProcess.emit (events.js:98:17)
    at maybeClose (child_process.js:756:16)
    at Process.ChildProcess._handle.onexit (child_process.js:823:5)
info: [debug] Responding to client with error: {"status":33,"value":{"message":"A new session could not be created. (Original error: android.util.AndroidException: INSTRUMENTATION_FAILED: com.getyourguide.android.selendroid/io.selendroid.ServerInstrumentation\r)","origValue":"android.util.AndroidException: INSTRUMENTATION_FAILED: com.getyourguide.android.selendroid/io.selendroid.ServerInstrumentation\r"},"sessionId":null}

```

**Comment by aluedeke** [comment]
WTF i finally figured out what went wrong:

The issue why instrumentation failed was because the selendroid apk, which was installed on the device, had the package name io.selendroid and target package io.selendroid.test (the original ones). So the issue is that the selendroid apk was not updated with the proper package names/manifest.xml for my app under test.

The reason that the selendroid apk had the wrong package names was 
1. it was never regenerated, (which i kind of don't like). So whatever i did i was impossible to fix the issue until i deleted the broken apk (/tmp/selendroid.packagename.apk)
2. after forcing appium to regenrate the apk it turned out the test zip archive step failed for the selendroid apk:

```
error: Test zip archive threw error Error: Command failed: 
error: Stderr: AndroidManifest.xml    bad extra-field entry:
      EF block length (0 bytes) invalid (< 4)
At least one error was detected in /tmp/selendroid.com.getyourguide.android.apk.

error: Stdout: 
```

After some time i stumbled upon bug report https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=775640
(which isn't fixed for debian and ubuntu based systems). So not an appium bug but an showstopper on (all?) recent linux based systems.

The bug in appium from my point of view is that the not finished/broken selendroid apk isn't deleted in this error case so that the user will never be able to figure out whats wrong.

I would prefer that the test zip step gets removed completely (its anyway not executed in windows) and that the selendroid apk gets always regenerated.

@bootstraponline @jlipps any thoughts?

**Comment by bootstraponline** [comment]
I think removing the [test zip step](https://github.com/appium/appium/blob/be56e12e49c78fb91ce0a236ac6680a6041264cd/lib/helpers.js#L101) is probably fine as a temp work around. We've had issues in the past with invalid zip files so it would be nice to detect them. I think relying on system binaries is bad as they can change. The node zip libraries are kind of terrible. Ideally we'd find some universal binary or library that we could bundle with appium to help with the zip related tasks.

**Comment by jlipps** [comment]
> I would prefer that the test zip step gets removed completely (its anyway not executed in windows) and that the selendroid apk gets always regenerated.

fine with me!

**Comment by paymand** [comment]
So this issue is a partly a dupe of https://github.com/appium/appium/issues/4215.
I mentioned the other part (unzip issue) on slack channel a while ago.
I'm also fine with removing the test step. The bug, however, is fixed and will be soon available on major distributions. On Ubuntu, it's not fixed yet as in version ``6.0-12ubuntu1.2``.
In the meantime, I have downgraded to working version on my test machines:
``sudo apt-get install unzip=6.0-12ubuntu1``

**Comment by paymand** [comment]
The latest unzip version on Ubuntu (6.0-12ubuntu1.3) has the fix and testZip step doesn't fail anymore.
